tempesta_fw/sock.c

```diff
@@ -483,7 +551,7 @@ __ss_close(struct sock *sk, int flags)
 	ss_do_close(sk);
 	bh_unlock_sock(sk);
 	if (flags & SS_F_CONN_CLOSE)
-		SS_CALL(connection_drop, sk);
+		SS_CALL_GUARD_EXIT(connection_drop, sk);
```

Comment:
> It means that if we're here in SS and we're doing a synchronous/immediate close(), then the socket is alive at this moment.

Yes. Tempesta layer can close a socket only if it has corresponding TfwConnection. When we link a socket with TfwConnectionn we increment the socket reference count in `tfw_connection_link_to_sk()`, so if Tempesta closes a socket, then it refers it and the socket is alive.

> Can it be called twice? I am not sure yet. Perhaps, that must be construed as a bug.

Yes, but we have reference counters which should protect the socket and TfwConnetion.